### PR TITLE
Operator for minimising the variance

### PR DESCRIPTION
**Pull request by alleSini99**
Operator for minimising the variance `<O^2> - <O>^2` of an arbitrary operator `O`. 
@PhilipVinc

### CODE REVIEW

**Comment by alleSini99** [comment]
@PhilipVinc

File: netket/experimental/observable/variance/variance_operator.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    def op(self) -> AbstractOperator:
+        """
+        The opeator for which the variance is to be computed.
+        """
+        return self._op
+
+    @property
+    def op2(self)-> AbstractOperator:
+        """
+        The squared of the opeator for which the variance is to be computed.
```
Comment (PhilipVinc) [inline netket/experimental/observable/variance/variance_operator.py:73]: Say that depending on the flag in the constructor this might be something or something else...

File: netket/experimental/observable/variance/variance_operator.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        super().__init__(op.hilbert)
+        self._op = op
+
+        if use_Oloc2:
+            self._op2 = nk.operator.Squared(op)
+        else:
+            self._op2 = op @ op
+
+    @property
+    def op(self) -> AbstractOperator:
```
Comment (PhilipVinc) [inline netket/experimental/observable/variance/variance_operator.py:64]: can we call those `òperator` instead of op and operator_squared instead of op2

File: netket/experimental/observable/variance/exact.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    vstate: FullSumState,
+    varop: VarianceOperator,
+    *,
+    mutable,
+):
+    if varop.hilbert != vstate.hilbert:
+        raise TypeError("Hilbert spaces should match")
+
+    op_mtrx = varop.op.to_dense()
+    op2_mtrx = varop.op2.to_dense()
```
Comment (PhilipVinc) [inline netket/experimental/observable/variance/exact.py:57]: as above

File: netket/experimental/observable/variance/exact.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        return O2_mean - O_mean**2
+
+    if not return_grad:
+        var = expect_kernel_var(params)
+        return Stats(mean=var, error_of_mean=0.0, variance=0.0)
+
+    var, var_vjp_fun = nkjax.vjp(expect_kernel_var, params, conjugate=True)
+
+    var_grad = var_vjp_fun(jnp.ones_like(var))[0]
+    var_grad = jax.tree_map(lambda x: mpi.mpi_mean_jax(x)[0], var_grad)
```
Comment (PhilipVinc) [inline netket/experimental/observable/variance/exact.py:92]: here you mean over mpi. maybe it's correct but can you double check? we should run the test under MPI

File: netket/experimental/observable/variance/exact.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        raise TypeError("Hilbert spaces should match")
+
+    op_mtrx = varop.op.to_dense()
+    op2_mtrx = varop.op2.to_dense()
+
+    return expect_and_grad_inner_fs(
+        vstate._apply_fun,
+        vstate.parameters,
+        vstate.model_state,
+        vstate._all_states,
```
Comment (PhilipVinc) [inline netket/experimental/observable/variance/exact.py:63]: all states are... all states on every rank?

File: netket/experimental/observable/variance/expect.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+                log_pdf,
+                lambda params, σ: local_kernel(
+                    afun, {"params": params, **model_state}, σ, args
+                ),
+                params,
+                σ,
+                n_chains=n_chains,
+            )[0]
+
+            O2_loc = local_kernel2(afun, W, σ, args2)
```
Comment (PhilipVinc) [inline netket/experimental/observable/variance/expect.py:120]: we should take the real part here? the variance should be real

File: netket/jax/_expect.py [hunk 7]
```diff
@@ -95,6 +96,8 @@ def f(pars, σ, *cost_args):
 
     _, pb = nkvjp(f, pars, σ, *cost_args)
     grad_f = pb(dL̄)
+    grad_f = jax.tree_map(lambda x: mpi_sum_jax(x)[0], grad_f)
```
Comment (PhilipVinc) [inline netket/jax/_expect.py:99]: porceddu baddu ohssignoreddu

Comment (gcarleo) [reply netket/jax/_expect.py:99]: I understand you are not happy about this line?

File: netket/experimental/observable/variance/variance_operator.py (hunk 1 above)
Comment (alleSini99) [reply netket/experimental/observable/variance/variance_operator.py]: Ok

File: netket/experimental/observable/variance/exact.py (hunk 3 above)
Comment (alleSini99) [reply netket/experimental/observable/variance/exact.py:57]: Yes

File: netket/experimental/observable/variance/expect.py (hunk 6 above)
Comment (alleSini99) [reply netket/experimental/observable/variance/expect.py:120]: True

File: netket/experimental/observable/variance/exact.py (hunk 5 above)
Comment (alleSini99) [reply netket/experimental/observable/variance/exact.py:63]: Yes

File: netket/jax/_expect.py (hunk 7 above)
Comment (alleSini99) [reply netket/jax/_expect.py:99]: This is in accordance with the merged PR #1356: we need a `mpi_sum_jax` as explained in the docstring/example in `nk.jax.expect`.

File: netket/experimental/observable/variance/exact.py (hunk 4 above)
Comment (alleSini99) [reply netket/experimental/observable/variance/exact.py:92]: I need a `mpi_sum_jax`, not `mpi_mean_jax`, as in `expect.py`, sorry. I am going to run the test under mpi to verify this.

File: netket/jax/_expect.py (hunk 7 above)
Comment (PhilipVinc) [reply netket/jax/_expect.py:99]: I mean, @alleSini99 , here you modified netket/jax/expect.py ?  I agree you should sum, but shouldn't you sum in the expect.py file of your operator?

Comment (PhilipVinc) [reply netket/jax/_expect.py:99]: (I think this is a stray edit from a previous version originating before PR 1356)

Comment (PhilipVinc) [reply netket/jax/_expect.py:99]: basically I think you should revert the change of this line?

File: test/observable/variance/test_variance.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
+import netket as nk
```
Comment (PhilipVinc) [inline test/observable/variance/test_variance.py:1]: just move this file inside of test/observable/test_variance.py , no need to create an extra folder.

**Review by PhilipVinc — COMMENTED** [review]
@alleSini99 looks good apart from those last two minor things

File: netket/jax/_expect.py (hunk 7 above)
Comment (alleSini99) [reply netket/jax/_expect.py:99]: No, yes, you are right, let me fix it.

File: test/observable/test_variance.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        vs = nk.vqs.MCState(
+            sampler=sa,
+            model=ma,
+            n_samples=n_samples,
+        )
+
+    else:
+        n_discard_per_chain = 1e3
+
+        sa = nk.sampler.MetropolisLocal(hilbert=hi, n_chains_per_rank_or_device=16)
```
Comment (PhilipVinc) [inline test/observable/test_variance.py:28]: I think something changed, because this should be n?chains per rank
```suggestion
        sa = nk.sampler.MetropolisLocal(hilbert=hi, n_chains_per_rank=16)
```

and same below

**Review by PhilipVinc — COMMENTED** [review]
@alleSini99 this last issue.

Also, can you add a line to the new features in the changelog?

**Review by PhilipVinc — COMMENTED** [review]
It's almost ok. 

- The formatting check is not passing (just run `pre-commit run --all-files` to fix it) 
- This test is not passing https://github.com/netket/netket/actions/runs/9839999768/job/27163249990?pr=1687#step:8:9451 maybe it's a numerical precision issue? Maybe just changing the seed might help. Or maybe it's areal problem (note that this happens when running tests under mPI)
- The documentation is not building but that is not your fault, but jax. I will fix that.

**Comment by alleSini99** [comment]
Indeed, the test under mpi was not passing for the `FullSumState` because in full summation we do the calculation without using `nk.jax.expect` but really with matrix and vector products. Therefore, we do not have to take into account that rescaling factor `n_nodes` present when using `nk.jax.expect` with mpi in the backward pass (see PR #1690). Therefore, we simply have to substitute the final `mpi_sum_jax` with an `mpi_mean_jax` (it is not necessary to sum the results on each rank since we are not dividing by a factor `n_nodes` in the ad, but simply averaging the result on each rank (that is the same for each rank)). In `expect`, instead, it is correct to do `mpi_sum_jax` since we use `nk.jax.expect` for the expectation value. Now, the tests pass both with and without mpi.

**Comment by PhilipVinc** [comment]
Still failing https://github.com/netket/netket/actions/runs/9874731068/job/27269795864?pr=1687#step:6:9542

**Review by PhilipVinc — APPROVED** [review]
